---
title: "cogena, a tool for co-expressed gene-set enrichment analysis and 
    visualization"
author: "Zhilong Jia, Michael Barnes"
date: "`r Sys.Date()`"
output:
#    rmarkdown::html_vignette:
#     word_document:
    pdf_document:
        fig_caption: yes
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\usepackage[utf8]{inputenc}
---

Co-expressed gene-set enrichment analysis, cogena, can be used as gene set 
enrichment analysis of co-expressed genes based on various clustering methods,
together with up and down regulated genes respectively. Gene sets could 
be Pathway, Gene Ontology, Cmap (Connectivity Map) compound or user-defined 
gene-sets. Accordingly, this tool can be used for pathway enrichment, 
GO enrichment, drug repositioning and so on. See the workflow
of cogena in Figure 1.


![Overview of the cogena workflow](Cogena_workflow.png)

## A quick start
See example in `?cogena` please.

## Input data
    Note: all the gene names should be gene SYMBOL since it is used in the 
    gene dataset files. Other kind of gene names could be used according to 
    the kind of gene names of user-defined gene-set files.

1. Input data needed.

- the gene expression profiling of differentially expressed genes. It can be 
matrix with gene in row and sample in column, data.frame or ExpressionSet 
object.
- the sample labels indicating the labels, like control and disease, of each 
sample. A vector with sample names.

2. Example dataset.

    There is an example data in `cogena` package. This dataset, from 
    [GSE20163](http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE20163), 
    is about Parkinson's disease. There are two objects in the PD dataset. 
    See `?PD` for more details.

```{r import_data, message=FALSE}
library(cogena)
data(PD)
# objects in the PD dataset.
```

```{r PD_data, echo=FALSE}
ls()
```

## Cogena analysis

1.  Gene sets

    There are a variety of gene sets in the `cogena` package, collected from 
    [MSigDB](http://www.broadinstitute.org/gsea/msigdb/index.jsp). They are 
    `r  dir(system.file("extdata", package="cogena"), pattern="*gmt")`. Here
    the gene-sets can be user-defined gene-sets formatted gmt, too. 
    Put it into the extdata directory in the installation directory of
    cogena, such as ~/R/x86_64-pc-linux-gnu-library/3.2/cogena/extdata in Linux). 

```{r gene_list_annotation}
annoGMT <- "c2.cp.kegg.v5.0.symbols.gmt"
annofile <- system.file("extdata", annoGMT, package="cogena")
```

2. Run cogena

    Setting some parameters for cogena and running cogena analysis.
```{r cogena, results="hide"}
# the number of clusters. It could be a vector.
# nClust <- 2:20
nClust <- 2:8
# the number of cores.
# ncore <- 8
ncore <- 2
# the clustering methods
# clMethods <- c("hierarchical","kmeans","diana","fanny","som","model",
# "sota","pam","clara","agnes") # All the methods can be used together.
clMethods <- c("hierarchical","kmeans")
# the distance metric
metric <- "correlation"
# the agglomeration method used for hierarchical clustering 
#(hierarchical and agnes)
method <- "complete"

# Clustering the genes
genecl_result <- coExp(DEexprs, nClust=nClust, clMethods=clMethods, 
                       metric=metric, method=method, ncore=ncore, verbose=TRUE)

#Enrichment analysis for the clusters
clen_res <- clEnrich(genecl_result, annofile=annofile, sampleLabel=sampleLabel)
```


## Analysing results of cogena
1.  After completing the cogena analysis, user can use `summary` to see 
the summary of the result of cogena. And `enrichment` caculate the enrichment 
score of certain clustering methods and certain number of cluster.

    Which clustering method and the number of clusters should be choose? Here 
    we show some principles:
    + Different clusters should account for different gene sets.
    + A gene set should enirched only in one cluster but not two or more.
    + The number of gene in a gene set enriched cluster should be as smaller as it can when the enrichment score is high.
    

```{r cogena_result_analysis}
summary(clen_res)
```



```{r enrichment}
# Here we consider the "hierarchical" method and 6 clusters.
# Always make the number as character, please!
enrichment.table <- enrichment(clen_res, "hierarchical", "6")
```

2.  The heatmap of expression profiling with clusters is plotted by 
    `heatmapCluster`. See Figure 2. Based on it, which cluster contains
    up-regulated or down-regulated genes is obvious.

```{r heatmapCluster, fig.width=10, fig.height=7, fig.cap="Heatmap of expression profiling with clusters"}
# Always make the number as character, please!
heatmapCluster(clen_res, "hierarchical", "6")
```

3.  To show the enrichment graph, `heatmapPEI` can be used. See Figure 3. 
    See `?heatmapPEI` for more details.

```{r heatmapPEI, fig.width=10, fig.height=6, fig.cap="KEGG pathway enrichment"}
# The enrichment score for 6 clusters, together with Down-regulated, 
# Up-regualted and All DE genes. The values shown in Figure 2 is the -log2(FDR).
#
# Always make the number as character, please!
heatmapPEI(clen_res, "hierarchical", "6", printGS=FALSE)

#ordered by the Cluster 1.
#heatmapPEI(cogena_result, "hierarchical", "6", orderMethod = "1")
```

4.  Genes and hub genes in a certain cluster. user can obtain the genes in 
    a certain cluster via `geneInCluster`, protein-protein interaction via
    `PPIinfo`, hubgenes in cluster via `hubGene`.
```{r geneInCluster}
# Always make the number as character, please!
geneC <- geneInCluster(clen_res, "hierarchical", "6", "2")
head(geneC)

```

5.  The gene expression profilings with cluster infromation could be 
    obtained by `geneExpInCluster`. There are two items, `clusterGeneExp` 
    and `label`, in the returned object of `geneExpInCluster`. This can 
    be used for other application.
```{r geneExpInCluster}
# Always make the number as character, please!
gec <- geneExpInCluster(clen_res, "hierarchical", "6")
head(gec$clusterGeneExp, 3)
gec$label
```

6. The correlation among one cluster could be shown via `corInCluster`. 
    see Figure 4.
```{r corInCluster, fig.width=6, fig.height=6, fig.cap="Correlation of genes in a cluster"}
# Always make the number as character, please!
corInCluster(clen_res, "hierarchical", "8", "8")
```

## Bug Report
https://github.com/zhilongjia/cogena/issues

## Citation
Jia Z. et al. *Cogena, a tool for co-expressed gene-set enrichment 
analysis and visualization*.

##Other Information
System info
```{r sessionInfo, echo=FALSE}
sessionInfo()
```
The END
