---
title: "cogena, a workflow for co-expressed gene-set enrichment analysis"
author: "Zhilong Jia, Michael R. Barnes"
date: "`r Sys.Date()`"
output:
    rmarkdown::html_vignette:
#     word_document:
#    pdf_document:
        fig_caption: yes
        number_sections: true
        css: custom.css
        toc: true
        toc_depth: 3
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\usepackage[utf8]{inputenc}
---

# Abstract

**Co**-expressed **g**ene-set **en**richment **a**nalysis, cogena, is a 
workflow for gene set enrichment analysis of co-expressed genes. The 
co-expression analysis is made based on various clustering methods. And The gene
set analysis can be pathway analysis, GO analysis, drug repositioning and so on.
The type of analysis cogena can is dependent on the gene set used in cogena. 
Lots kinds of gene sets are integrated into cogena from Msigdb and Connectivity
Map. See the workflow of cogena in Figure 1.


![Overview of the cogena workflow](Cogena_workflow.png)

# A quick start
See examples in `?cogena` please.

# Data Input
    Note: all the gene names should be gene SYMBOLs since they are used in the 
    gene set files. Other kind of gene names can be used according to 
    the kind of gene names of user-defined gene-set files.

## Input data required

- the gene expression profiling of differentially expressed genes. It can be 
matrix with gene in row and sample in column, data.frame or ExpressionSet 
object.
- the sample labels indicating the labels, like control and disease, of each 
sample. A vector with sample names.

## Example dataset

There is an example dataset in `cogena` package. This dataset, from 
[GSE13355](http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE13355), 
is about psoriasis. There are two objects in the Psoriasis dataset. 
See `?Psoriasis` for more details.

```{r import_data, message=FALSE}
library(cogena)
data(Psoriasis)
# objects in the Psoriasis dataset.
# Note: lable of interest should be behind of control label as this will affect
# the direction of gene regulation. For instance.
# Use factor(c("Normal", "Cancer", "Normal"), levels=c("Normal", "Cancer")), 
# instead of factor(c("Normal", "Cancer","Normal")) since "Cancer" is the 
# lable of interest.

```

```{r PD_data, echo=FALSE}
ls()
```

# Various Analyses

##  What kind of analysis can be done?

**The gene set used determines the type of analysis.**

There are a variety of gene sets in the `cogena` package, partly collected 
from [MSigDB](http://www.broadinstitute.org/gsea/msigdb/index.jsp) 
and [CMap](connectivitymap.org/cmap). They are 
_**`r  dir(system.file("extdata", package="cogena"), pattern="*gmt*")`**_. Here
the gene-sets can be user-defined gene-sets formatted gmt and compressed by
xz, too (such as c2.cp.kegg.v5.0.symbols.gmt.xz). Copy it into the extdata 
directory in the installation directory of cogena, such as
~/R/x86_64-pc-linux-gnu-library/3.2/cogena/extdata in Linux),
or kindly send to the author of cogena to share with others. 

## Types of analyses
* Pahtway Analysis
* GO Analysis
* Drug repositioning
* ...

# Pathway Analysis

KEGG Pathway Analysis, as an example, will be done to show the utility of
cogena.

## Parameter setting
```{r gene_list_annotation}
# KEGG Pathway gene set
annoGMT <- "c2.cp.kegg.v5.0.symbols.gmt.xz"
# GO biological process gene set
# annoGMT <- "c5.bp.v5.0.symbols.gmt.xz"
annofile <- system.file("extdata", annoGMT, package="cogena")

# the number of clusters. It can be a vector.
# nClust <- 2:20
nClust <- 10

# the number of cores.
# ncore <- 8
ncore <- 2

# the clustering methods
# clMethods <- c("hierarchical","kmeans","diana","fanny","som","model",
# "sota","pam","clara","agnes") # All the methods can be used together.
clMethods <- c("hierarchical","pam")

# the distance metric
metric <- "correlation"

# the agglomeration method used for hierarchical clustering 
# (hierarchical and agnes)
method <- "complete"
```

## Cogena running

```{r cogena, results="hide"}
# Co-expression Analysis
genecl_result <- coExp(DEexprs, nClust=nClust, clMethods=clMethods, 
                       metric=metric, method=method, ncore=ncore, verbose=TRUE)

# Enrichment (Pathway) analysis for the co-expressed genes
clen_res <- clEnrich(genecl_result, annofile=annofile, sampleLabel=sampleLabel)
```


## Results of pathway analysis
###  Summary of cogena result

After completing the cogena analysis, user can use `summary` to see 
the summary of the result of cogena. And `enrichment` caculate the enrichment 
score of certain clustering methods and certain number of cluster.

Which clustering method and the number of cluster should be choose? There is
no an automatical way. Here we show some principles:

+ Different clusters should account for different gene sets.
+ A gene set should enirched only in one cluster but not two or more.
+ The number of gene in a gene set enriched cluster should be as smaller as it can when the enrichment score is high.
    

```{r cogena_result_analysis}
summary(clen_res)
```

```{r enrichment}
# Here we consider the "pam" method and 10 clusters.
# Always make the number as character, please!
enrichment.table <- enrichment(clen_res, "pam", "10")
```

###  Heatmap of expression profiling with clusters
    
`heatmapCluster` is developed to show it. Based on Figure 2, it is obvious to know 
which cluster contains up-regulated or down-regulated genes.

```{r heatmapCluster, fig.width=10, fig.height=7, fig.cap="Heatmap of expression profiling with clusters"}
# Always make the number as character, please!
heatmapCluster(clen_res, "pam", "10", maintitle="Psoriasis")
```

### Enrichment heatmap of co-expressed genes 
    
`heatmapPEI` can be used to show the enrichment graph. See Figure 3. 
See `?heatmapPEI` for more details.

```{r heatmapPEI, fig.width=11, fig.height=7, fig.cap="KEGG pathway enrichment"}
# The enrichment score for 10 clusters, together with Down-regulated, 
# Up-regualted and All DE genes. The values shown in Figure 2 is the -log2(FDR).
#
# Always make the number as character, please!
heatmapPEI(clen_res, "pam", "10", printGS=FALSE, maintitle="Pathway analysis for Psoriasis")

```

##  Others

### Querying genes in a certain cluster

User can obtain the genes in a certain cluster via `geneInCluster`.
```{r geneInCluster}
# Always make the number as character, please!
geneC <- geneInCluster(clen_res, "pam", "10", "4")
head(geneC)

```

###  Gene expression profiling with cluster infromation

It can be obtained by `geneExpInCluster`. There are two items, 
`clusterGeneExp` and `label`, in the returned object of `geneExpInCluster`. 
It can be used for other application.
```{r geneExpInCluster}
# Always make the number as character, please!
gec <- geneExpInCluster(clen_res, "pam", "10")
gec$clusterGeneExp[1:3, 1:4]
gec$label[1:4]
```

### The gene correlation in a cluster

The correlation among a cluster can be checked and shown by `corInCluster`. See Figure 4.
```{r corInCluster, fig.width=6, fig.height=6, fig.cap="Correlation of genes in a cluster"}
# Always make the number as character, please!
corInCluster(clen_res, "pam", "10", "10")
```

# Drug repositioning

Drug repositioning is made based on co-expressed genes instead of all the 
differentially expressed genes. If the input of cogena is a disease related 
data, the drugs enriched should recovery the gene expression changed by the
disease, while if the input is drug related, the enriched drugs should show 
similar gene expression changes caused by the drug studied. Here we show drugs
for treating psoriasis, an autoimmune disease.

## Drug repositioning analysis running

The drug repositioning gene set choice of _CmapDn100.gmt.xz_ or 
_CmapUp100.gmt.xz_ should be made based on
the regulation direction of clusters. For example, as the 7th cluster contains 
up-regulated genes for psoriasis, the _CmapDn100.gmt.xz_ is choosen for drug 
repositioning of psoriasis to recovery the gene expression changes caused by 
the disease.

```{r drp}
# A comprehensive way
# cmapDn100_cogena_result <- clEnrich(genecl_result, 
# annofile=system.file("extdata", "CmapDn100.gmt.xz", package="cogena"), 
# sampleLabel=sampleLabel)

# A quick way
# Based on the pathway analysis results
cmapDn100_cogena_result <- clEnrich_one(genecl_result, "pam", "10", 
    annofile=system.file("extdata", "CmapDn100.gmt.xz", package="cogena"), 
    sampleLabel=sampleLabel)
```

## Original result of drug repositioning
    Showing the results ordered by the 7th cluster in Figure 5.
```{r drp_figure, fig.width=10, fig.height=7, fig.cap="Drug repositioning"}
heatmapPEI(cmapDn100_cogena_result, "pam", "10", printGS=FALSE, 
           orderMethod = "7", maintitle="Drug repositioning for Psoriasis")

# Results based on cluster 5.
# heatmapPEI(cmapDn100_cogena_result, "pam", "10", printGS=FALSE, 
#           orderMethod = "5", maintitle="Drug repositioning for Psoriasis")

# Results based on cluster 9, containing down-regulated genes.
# heatmapPEI(cmapUp100_cogena_result, "pam", "10", printGS=FALSE, 
#           orderMethod = "9", maintitle="Drug repositioning for Psoriasis")

```

## Multi-instance merged result of drug repositioning

Usually there are more than 1 instances for a drug with different doses
or time-points in the Cmap gene set. `heatmapCmap` can merge the multi-instance
results based on parameter _mergeMethod_ ("mean" or "max"). Figure 6 shows the multi-instance merged results ordered by the 7th cluster.
```{r drp_figure2, fig.width=10, fig.height=7, fig.cap="Drug repositioning (multi-instance merged)"}
heatmapCmap(cmapDn100_cogena_result, "pam", "10", printGS=FALSE, 
            orderMethod = "7", maintitle="Drug repositioning for Psoriasis")
```


# Bug Report

https://github.com/zhilongjia/cogena/issues

# Citation

Jia Z. et al. *Cogena, a tool for co-expressed gene-set enrichment 
analysis and visualization*.

# Other Information

System info
```{r sessionInfo, echo=FALSE}
sessionInfo()
```

